src/common/pf_bg_worker.c

```diff
@@ -0,0 +1,117 @@
+ *        _       _         _
+ *  _ __ | |_  _ | |  __ _ | |__   ___
+ * | '__|| __|(_)| | / _` || '_ \ / __|
+ * | |   | |_  _ | || (_| || |_) |\__ \
+ * |_|    \__|(_)|_| \__,_||_.__/ |___/
+ *
+ * www.rt-labs.com
+ * Copyright 2032 rt-labs AB, Sweden.
+ *
+ * This software is dual-licensed under GPLv3 and a commercial
+ * license. See the file LICENSE.md distributed with this software for
+ * full license information.
+ ********************************************************************/
+
+#include "pf_includes.h"
+
+#include <inttypes.h>
+
+/* Events handled by bg worker task */
+#define BG_JOB_EVENT_SAVE_NVM_DATA       BIT (0)
+#define BG_JOB_EVENT_UPDATE_PORTS_STATUS BIT (1)
+
+static void bg_worker_task (void * arg);
+static void run_save_file_job (pnet_t * net);
+static void run_update_ports_status_job (pnet_t * net);
+
+void pf_bg_worker_init (pnet_t * net)
+{
+   net->pf_bg_worker.events = os_event_create();
+   CC_ASSERT (net->pf_bg_worker.events != NULL);
+
+   os_thread_create (
+      "p-net_bg_worker",
+      10, // todo
+      4000,
+      bg_worker_task,
+      (void *)net);
+}
+
+int pf_bg_worker_run_job (pnet_t * net, pf_bg_job_t job_id)
+{
+   if (net == NULL)
+   {
+      printf ("pf_bg_worker_run_job net==NULL\n");
+      return -1;
+   }
+
+   switch (job_id)
+   {
+   case BG_JOB_SAVE_NVM_DATA:
+      os_event_set (net->pf_bg_worker.events, BG_JOB_EVENT_SAVE_NVM_DATA);
+      break;
+   case BG_JOB_UPDATE_PORTS_STATUS:
+      os_event_set (net->pf_bg_worker.events, BG_JOB_EVENT_UPDATE_PORTS_STATUS);
+      break;
+   default:
+      LOG_ERROR (
+         PNET_LOG,
+         "BGW(%d): Unsupported job %" PRIu32 "\n",
+         __LINE__,
+         job_id);
+      return -1;
+   }
+
+   return 0;
+}
+
+static void bg_worker_task (void * arg)
+{
+   pnet_t * net_handle = (pnet_t *)arg;
+   uint32_t mask = BG_JOB_EVENT_SAVE_NVM_DATA |
+                   BG_JOB_EVENT_UPDATE_PORTS_STATUS;
+   uint32_t flags = 0;
+
+   for (;;)
+   {
+      os_event_wait (
+         net_handle->pf_bg_worker.events,
+         mask,
+         &flags,
+         OS_WAIT_FOREVER);
+
+      if (flags & BG_JOB_EVENT_SAVE_NVM_DATA)
+      {
+         os_event_clr (
+            net_handle->pf_bg_worker.events,
+            BG_JOB_EVENT_SAVE_NVM_DATA);
+         {
+            run_save_file_job (net_handle);
+         }
+      }
+      if (flags & BG_JOB_EVENT_UPDATE_PORTS_STATUS)
+      {
+         os_event_clr (
+            net_handle->pf_bg_worker.events,
+            BG_JOB_EVENT_UPDATE_PORTS_STATUS);
+         {
+            run_update_ports_status_job (net_handle);
+         }
+      }
+   }
+}
+
+static void run_save_file_job (pnet_t * net)
+{
+   LOG_DEBUG (PNET_LOG, "BGW(%d): Running save file job\n", __LINE__);
+   pf_cmina_save_ase (net, &net->cmina_nonvolatile_dcp_ase);
+   pf_fspm_save_im (net);
+   pf_pdport_save_all (net);
+}
+
+static void run_update_ports_status_job (pnet_t * net)
```

Comment:
```
/**
 * Run a job updating the status for all physical ports.
 *
 * @param net              InOut: The p-net stack instance
 */
```

Reply:
Removed the inernal functions and instead calling the functions doing the job directly in the event handler loop. I think this is a better solution after splitting up the save job to different jobs (IM,ASE,PDPORT).
OK